### PR TITLE
fix(auth): identifier-status returned exists=false for real users (forced OTP on every login)

### PR DESCRIPTION
**Pull request by RaviSahu1520**
## Summary
Fixes the login bug where a **returning, verified user with a password** was forced through OTP + Set Password on every login instead of being offered a password login.

## Root cause
`POST /api/v1/auth/identifier-status` looked the user up via **GoTrue's Admin API** (`admin_find_user_by_phone` / `admin_get_user_by_email`). That lookup returned **no match for real users — phone and email, even exact stored values** — because GoTrue's admin `/admin/users` endpoint does not filter by the `phone`/`email` query params the code passed, and the old code silently treated any non-match / API hiccup as "user not found".

Result: `exists=false` → `next_step="otp"` for everyone → the client sent every returning user back through OTP + Set Password.

Verified by running the old functions against real users:
```
admin_find_user_by_phone('919310833204')              -> NONE   # exact stored value
admin_find_user_by_phone('+919310833204')             -> NONE
admin_get_user_by_email('ankit.mishra21@gmail.com')   -> NONE   # exact stored value
```

## Fix
Replace the Admin API lookup with a **direct, authoritative query against `auth.users`** (one read):
- `exists` — a row matched
- `verified` — `email_confirmed_at` / `phone_confirmed_at`
- `has_password` — `encrypted_password IS NOT NULL`
- Phone matches both E.164 (`+9…`) and the GoTrue-stored form (no leading `+`); email is case-insensitive.
- Any DB error raises **503** instead of silently treating an existing user as new.

> **Note on `has_password`:** this PR also switches that field from the `app_metadata.providers` heuristic to `encrypted_password`. That part is a **robustness/correctness improvement, not the bug fix** — the old heuristic agreed with `encrypted_password` for 118/119 real users. The essential fix is the **lookup method** change. `encrypted_password` is kept because it's the authoritative signal (the `providers` proxy conflates "has the email/phone channel" with "has a password").

## Testing
- `TestGetIdentifierStatus` rewritten — 12 cases (verified+password→`password`, verified no-password→`otp`, unverified→`otp`, unknown→`otp`, DB-error→`ServiceUnavailableException`, phone E.164 normalization ×4, email case-insensitivity). API-layer tests unchanged (they mock the whole function).
- `uv run ruff check app/` clean.
- Verified **live against the real DB**: confirmed phone+password → `next_step="password"`; confirmed email+password → `password`; unknown identifier → `otp`.

## Files changed
- `app/services/user.py` — rewrote `get_identifier_status` (direct `auth.users` query) + `_normalize_phone_to_e164` helper.
- `app/api/api_v1/endpoints/auth.py` — added `db: AsyncSession = Depends(get_db)` to the endpoint.
- `tests/unit/services/test_user_service.py` — rewrote `TestGetIdentifierStatus`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://stagereview.app/360ghar/backend/pull/18">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

---
## Summary by cubic
Fixes a login bug that forced returning, verified users with passwords into OTP + Set Password on every login. `POST /api/v1/auth/identifier-status` now correctly detects existing users and offers password login when appropriate.

- **Bug Fixes**
  - Replaced GoTrue Admin lookup with a direct `auth.users` query to compute `exists`, `verified`, and `has_password` (`encrypted_password IS NOT NULL`).
  - Phone is normalized to E.164 and matched with and without the leading “+” (`removeprefix("+")`); email match is case-insensitive and index-safe (`email = lower(:email)`).
  - DB lookup failures raise 503; only the exception type is logged to avoid leaking identifiers. Phone normalization and query construction now happen inside the try block.
  - Endpoint injects `db: AsyncSession` and calls `get_identifier_status(db, identifier)`.

<sup>Written for commit 9abfe9f27b8175437a23c0699624b6b49f0d2c6b. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/360ghar/backend/pull/18?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

<h3>Greptile Summary</h3>

Replaces a broken GoTrue Admin API lookup (which silently returned `None` for real users) with a single direct `SELECT` against `auth.users`, so returning users with a password are correctly offered password login instead of being forced through OTP on every attempt.

- **`app/services/user.py`**: `get_identifier_status` now takes an `AsyncSession`, constructs a `text()` SQL query against `auth.users` for `email_confirmed_at`, `phone_confirmed_at`, and `encrypted_password IS NOT NULL`; phone matching covers both E.164 (`+91…`) and GoTrue's stored form (`91…`); DB failures raise `ServiceUnavailableException` (503) instead of masking users as new.
- **`app/api/api_v1/endpoints/auth.py`**: `db: AsyncSession = Depends(get_db)` injected and forwarded to the service; no other logic changed.
- **`tests/unit/services/test_user_service.py`**: 12-case test class replaces the old mock-based suite, covering verified/unverified, password/no-password, unknown identifiers, DB-error-to-503, E.164 normalization, and email case-insensitivity.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the direct `auth.users` query is authoritative, DB errors are explicitly surfaced as 503, and the fix is covered by 12 unit tests that were verified against the live database.

The change is narrowly scoped: one service function is rewritten, one endpoint gains a DB dependency, and the test suite is fully replaced. The SQL query is read-only against a single table, the error path is deterministic (any exception → 503, not silent exists=false), and both phone normalization edge cases and email case-insensitivity are explicitly tested. No other code paths are touched.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| app/services/user.py | Core fix: rewrites `get_identifier_status` to query `auth.users` directly via SQLAlchemy `text()` instead of the broken GoTrue Admin API; adds `_normalize_phone_to_e164` helper; raises 503 on DB errors instead of silently returning `exists=false`. |
| app/api/api_v1/endpoints/auth.py | Minimal, correct change: injects `db: AsyncSession = Depends(get_db)` and threads it through to `get_identifier_status(db, identifier)`; docstring updated to reflect the new lookup mechanism. |
| tests/unit/services/test_user_service.py | Test class fully rewritten with 12 cases covering verified/unverified email+phone, password vs no-password, unknown identifiers, DB error → 503, phone E.164 normalization (×4 parametrized), and email stripping; all assertions align correctly with the new implementation. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Client
    participant Endpoint as POST /auth/identifier-status
    participant Service as get_identifier_status()
    participant DB as auth.users (Postgres)

    Client->>Endpoint: identifier (email or phone)
    Endpoint->>Endpoint: rate-limit check
    Endpoint->>Service: get_identifier_status(db, identifier)
    alt email channel
        Service->>DB: "SELECT … WHERE email = lower(:email)"
    else phone channel
        Service->>Service: _normalize_phone_to_e164(identifier)
        Service->>DB: SELECT … WHERE phone IN (:phone_e164, :phone_noplus)
    end
    alt DB error
        DB-->>Service: exception
        Service-->>Endpoint: ServiceUnavailableException (503)
    else no row found
        DB-->>Service: None
        Service-->>Endpoint: "{exists:false, next_step:otp}"
    else row found
        DB-->>Service: "{email_confirmed_at, phone_confirmed_at, has_password}"
        Service-->>Endpoint: "{exists, verified, has_password, next_step}"
    end
    Endpoint-->>Client: IdentifierStatusResponse
```

<sub>Reviews (2): Last reviewed commit: ["fix(auth): address PR review feedback"](https://github.com/360ghar/backend/commit/9abfe9f27b8175437a23c0699624b6b49f0d2c6b) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37400430)</sub>

### CODE REVIEW

**Comment by ghost** [comment]
Ready to review this PR? Stage has broken it down into 4 individual chapters for you:

| | Title |
|---|-------|
| 1 | [Update imports and exceptions for user service](https://stagereview.app/360ghar/backend/pull/18/chapters/1) |
| 2 | [Rewrite identifier status lookup with direct DB query](https://stagereview.app/360ghar/backend/pull/18/chapters/2) |
| 3 | [Wire database session into auth endpoint](https://stagereview.app/360ghar/backend/pull/18/chapters/3) |
| 4 | [Update unit tests for direct DB lookup](https://stagereview.app/360ghar/backend/pull/18/chapters/4) |

<a href="https://stagereview.app/360ghar/backend/pull/18">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<sub>Chapters generated by <a href="https://stagereview.app">Stage</a> for commit 9abfe9f27b8175437a23c0699624b6b49f0d2c6b on Jun 13, 2026 1:32pm UTC.</sub>

File: tests/unit/services/test_user_service.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 """
 
 import uuid
-from unittest.mock import AsyncMock, patch
+from unittest.mock import AsyncMock, MagicMock, patch
```
[the remark this answers is not in the document]
Comment (RaviSahu1520) [reply tests/unit/services/test_user_service.py:8]: ✅ Fixed in 9abfe9f — added `from __future__ import annotations` as the first import in both `app/services/user.py` and `tests/unit/services/test_user_service.py`.

File: app/services/user.py [hunk 2]
```diff
@@ -306,65 +307,100 @@ async def set_last_auth_method(db: AsyncSession, user: User, method: AuthMethod)
...
-            "verified": False,
-            "has_password": False,
-            "channel": channel,
-            "next_step": "otp",
-        }
-
-    exists = record is not None
+        raise ServiceUnavailableException(
+            detail="Identity provider is temporarily unavailable, please retry",
+        ) from exc
```
[the remark this answers is not in the document]
Comment (RaviSahu1520) [reply app/services/user.py:402]: ✅ Fixed in 9abfe9f — the failure log now records only the exception type (`type(exc).__name__`), not the exception message, so SQL/bound params (email/phone) can’t leak via DBAPI error strings on this public endpoint.

File: app/services/user.py [hunk 3]
```diff
@@ -306,65 +307,100 @@ async def set_last_auth_method(db: AsyncSession, user: User, method: AuthMethod)
...
+    try:
+        result = await db.execute(stmt, params)
+        row = result.mappings().first()
+    except Exception as exc:  # noqa: BLE001 — any DB failure must not misroute
         logger.warning(
-            "identifier-status: Supabase lookup failed for %s channel; "
-            "returning exists=false",
+            "identifier-status: auth.users lookup failed for %s channel: %s",
             channel,
+            exc,
```
[the remark this answers is not in the document]
Comment (RaviSahu1520) [reply app/services/user.py]: ✅ Fixed in 9abfe9f — the failure log now records only the exception type (`type(exc).__name__`), not the exception message, so SQL/bound params (email/phone) can’t leak via DBAPI error strings on this public endpoint.

File: app/services/user.py [hunk 4]
```diff
@@ -306,65 +307,100 @@ async def set_last_auth_method(db: AsyncSession, user: User, method: AuthMethod)
...
-        record = await _manager.admin_find_user_by_phone(identifier)
-
-    # Handle Supabase provider failures — tagged failure dicts from
-    # _make_failure should not be treated as real user records.
-    from app.core.auth import _is_failure
+        # GoTrue stores auth.users.phone WITHOUT a leading "+" (e.g.
+        # "916280137577"); the normalized E.164 form has it ("+916280137577").
+        # Match both forms so the lookup is robust to either storage style.
+        phone_value = _normalize_phone_to_e164(identifier)
+        phone_noplus = phone_value.lstrip("+")
```
[the remark this answers is not in the document]
Comment (RaviSahu1520) [reply app/services/user.py]: ✅ Fixed in 9abfe9f — switched to `removeprefix('+')` so at most one leading '+' is stripped (malformed input can’t be massaged into a valid key).

File: tests/unit/services/test_user_service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (RaviSahu1520) [reply tests/unit/services/test_user_service.py:8]: ✅ Fixed in 9abfe9f — added `from __future__ import annotations` as the first import in both `app/services/user.py` and `tests/unit/services/test_user_service.py`.

File: app/services/user.py [hunk 5]
```diff
@@ -1,14 +1,15 @@
 from typing import Any
 
-from sqlalchemy import and_, func, or_, select
+from sqlalchemy import and_, func, or_, select, text
```
[the remark this answers is not in the document]
Comment (RaviSahu1520) [reply app/services/user.py:5]: ✅ Fixed in 9abfe9f — added `from __future__ import annotations` as the first import in both `app/services/user.py` and `tests/unit/services/test_user_service.py`.

File: app/services/user.py [hunk 6]
```diff
@@ -306,65 +307,100 @@ async def set_last_auth_method(db: AsyncSession, user: User, method: AuthMethod)
...
+            """
+            SELECT id, email, phone,
+                   email_confirmed_at, phone_confirmed_at,
+                   (encrypted_password IS NOT NULL) AS has_password
+            FROM auth.users
+            WHERE phone IN (:phone_e164, :phone_noplus)
+            LIMIT 1
+            """
+        )
+        params = {"phone_e164": phone_value, "phone_noplus": phone_noplus}
```
[the remark this answers is not in the document]
Comment (RaviSahu1520) [reply app/services/user.py]: ✅ Fixed in 9abfe9f — phone normalization and query construction are now inside the `try` block, so any failure surfaces as 503 per the docstring contract.

File: app/services/user.py [hunk 7]
```diff
@@ -306,65 +307,100 @@ async def set_last_auth_method(db: AsyncSession, user: User, method: AuthMethod)
...
 
     if channel == "email":
-        record = await _manager.admin_get_user_by_email(identifier)
+        stmt = text(
+            """
+            SELECT id, email, phone,
+                   email_confirmed_at, phone_confirmed_at,
+                   (encrypted_password IS NOT NULL) AS has_password
+            FROM auth.users
+            WHERE lower(email) = lower(:email)
```
[the remark this answers is not in the document]
Comment (RaviSahu1520) [reply app/services/user.py]: ✅ Fixed in 9abfe9f — changed to `email = lower(:email)` (lower on the bound value, not the column) so the unique btree index on `auth.users.email` stays usable; correct because GoTrue stores emails lowercased.

**Review by saksham1991999 — APPROVED** [review]
